### PR TITLE
Table align cleanup

### PR DESCRIPTION
**Pull request by taldcroft**
This fixes up #3644 by cleaning up the code logic, fixing docstrings, and making the UI more convenient and consistent.  It should also be a lot faster by moving code out of the formatting inner loop.  It also adds a number of tests.

### CODE REVIEW

**Comment by pllim** [comment]
:+1:

File: astropy/table/pprint.py [hunk 1]
```diff
@@ -482,32 +459,32 @@ def _pformat_table(self, table, max_lines=None, max_width=None,
...
+        # Coerce align into a correctly-sized list of alignments (if possible)
+        n_cols = len(table.columns)
+        if align is None:
+            align = [None] * n_cols
+
+        if isinstance(align, six.string_types):
+            align = [align]
+
+        if isinstance(align, (list, tuple)):
+            if len(align) == 1:
```
Comment (taldcroft) [inline astropy/table/pprint.py]: When I think about it I'm not totally convinced we should coerce a one-element list to the right size.  This could just be specified with a single string.  Question is whether this would allow a code error to go unnoticed.

File: astropy/table/pprint.py [hunk 2]
```diff
@@ -266,51 +266,27 @@ def _pformat_col(self, col, max_lines=None, show_name=True, show_unit=None,
...
+            # Format columns according to alignment.  `align` arg has precedent, otherwise
+            # use `col.format` if it is a legal alignment string.  If neither applies
+            # then right justify.
+            justify_methods = {'<': 'ljust', '^': 'center', '>': 'rjust', '0=': 'zfill'}
+            align = align or (col.info.format
+                              if col.info.format in justify_methods
+                              else '>')
+
+            # This can only occur if `align` was explicitly provided.
+            if align not in justify_methods:
```
Comment (mhvk) [inline astropy/table/pprint.py]: @taldcroft - this leaves much less freedom to use the alignment than the admittedly long if-else clause that you removed, in that previously formats like `>8.3f` would work, as would `.<20.3f` (which would put dots to fill up any space). E.g., with the implementation as is one can get ApJ-table style effects as follows:
```
In [5]: t = Table([['src 1', 'src 2 with a long name']], names=['a'])

In [6]: t['a'].format = '.<'

In [7]: t
Out[7]: 
<Table length=2>
          a           
        str22         
----------------------
src 1.................
src 2 with a long name
```

Can we keep the long statement but just move it out of the loop? (There is no reason for it to be inside.) Though for the case that `align` is defined, the big construction is needless; don't know how that passed by me...
```
    else:
        justify_method = justify_methods.get(align,

File: astropy/table/pprint.py [hunk 3]
```diff
@@ -482,32 +459,32 @@ def _pformat_table(self, table, max_lines=None, max_width=None,
...
-                                                align=align[i])
-                if outs['show_length']:
-                    lines = lines[:-1]
-                cols.append(lines)
+        # Coerce align into a correctly-sized list of alignments (if possible)
+        n_cols = len(table.columns)
+        if align is None:
+            align = [None] * n_cols
+
+        if isinstance(align, six.string_types):
```
Comment (mhvk) [inline astropy/table/pprint.py]: I would just do `elif` and set to `[align] * n_cols`

File: astropy/table/pprint.py [hunk 4]
```diff
@@ -482,32 +459,32 @@ def _pformat_table(self, table, max_lines=None, max_width=None,
...
+            align = [align]
+
+        if isinstance(align, (list, tuple)):
```
Comment (mhvk) [inline astropy/table/pprint.py]: and another `elif` here.

File: astropy/table/pprint.py (hunk 1 above)
Comment (mhvk) [inline astropy/table/pprint.py]: Agreed. Seems an unnecessary exception.

File: astropy/table/pprint.py (hunk 3 above)
Comment (mhvk) [inline astropy/table/pprint.py]: For this case where one gives an alignment that holds for all columns, I feel it would be best if it did *not* override any alignment given in the column format. But probably that is a bit too tricky to implement...

File: astropy/table/pprint.py (hunk 2 above)
Comment (taldcroft) [inline astropy/table/pprint.py]: I'll think about it, but part of the problem was that this mini-language was entirely undocumented and unvalidated.  I should have refreshed myself with the original thread, but coming at it fresh I didn't even know what was intended by the code (and it looks fragile, like expected `col.format[1]` to exist, though maybe that can't really be a problem?).  

Overall I was just worried by the interactions between the `_pformat_col_iter` processing and this.

Comment (taldcroft) [inline astropy/table/pprint.py]: Ah wait, I think I get it.

**Comment by taldcroft** [comment]
@mhvk - I put back fill characters using a completely different implementation.  It's only a little bit shorter, but it now I can look at the code and easily convince myself it should work.

**Comment by taldcroft** [comment]
I'm hoping the narrative docs pass tests.  In my local environment it still isn't running those tests.

**Comment by taldcroft** [comment]
Tests passing, ready for final review.

**Comment by mhvk** [comment]
@taldcroft - even for me my implementation had become hard to parse, and I think yours is much clearer!  And the tests passed, so merging.
